heatmap/src/heatmap.rs

```diff
@@ -196,40 +209,52 @@ impl Heatmap {
         self.summary.buckets()
     }
 
+    /// Returns the number of valid and active `Histogram` slices in the `Heatmap`
+    pub fn active_slices(&self) -> usize {
+        if Instant::now() - self.first_tick > self.span {
+            self.slices()
+        } else {
+            self.idx.load(Ordering::Relaxed) + 1
+        }
+    }
+
+    /// Returns the DateTime representation of when the `HeatMapi` was created
+    pub fn created_at(&self) -> DateTime {
+        self.create_at
+    }
+
     /// Increment a time-value pair by a specified count
     pub fn increment(&self, time: Instant, value: u64, count: u32) {
         self.tick(time);
 
-        let current = self.current.load(Ordering::Relaxed);
+        let curr_tick = self.curr_tick.load(Ordering::Relaxed);
+        let idx = self.idx.load(Ordering::Relaxed);
 
         // fast path when the time falls into the current time slice
-        if time >= self.slices[current].start.load(Ordering::Relaxed) {
+        if time >= curr_tick {
             let _ = self.summary.increment(value, count);
-            let _ = self.slices[current].histogram.increment(value, count);
+            let _ = self.histograms[idx].increment(value, count);
         }
 
-        // since it's before the current time slice, we need to record it into
-        // a past slice
+        // the duration belonged to a past histogram
 
-        let start = self.start.load(Ordering::Relaxed);
-        let stop = self.stop.load(Ordering::Relaxed);
+        // first we calculated much before current tick the event happened
+        let behind = curr_tick.duration_since(time);
+        let idx_backward = (behind.as_nanos() / self.resolution.as_nanos()) as usize;
 
-        // if the sample is before the heatmap start, we drop it
-        if time < start {
+        if idx_backward > self.slices() {
+            // We may want to log something ehre
```

Comment:
Returning an error here would be most correct. We already have errors if the value itself is out of range. We should add another Error type for the timestamp being out of range and return that here. Fine addressing in a follow-up PR though.

Reply:
Alternatively - update the description for `Error::OutOfRange` and repurpose that to indicate that the time-value pair is out of range in one or both dimmensions

Reply:
Oh good point. I should add an error type, probably its own since I personally prefer not to mix dimensions :)